modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/FDSView.tsx

```diff
@@ -13,10 +13,10 @@ import {API_URL, OBJECT_RELATIONSHIP} from './Constants';
 import {FDSViewType} from './FDSViews';
 import Actions from './fds_view/Actions';
 import Details from './fds_view/Details';
-import Fields from './fds_view/Fields';
 import Filters from './fds_view/Filters';
 import Pagination from './fds_view/Pagination';
 import Sorting from './fds_view/Sorting';
+import Fields from './fds_view/fields/Fields';
```

Comment:
@juanjofgliferay we recently had a discussion about directory structures in review comments, but I can't find it. I'm not a fan of our current solutions in DSM, and they are inconsistent. The pattern I'm suggesting in PR is a directory named the same as subcomponent, with everything related to it inside. The goal is to make it easier to find "everything related to fields".

Reply:
Previous discussion -> https://github.com/liferay-frontend/liferay-portal/pull/3727#discussion_r1352637791

I like the approach, having a folder for each main component (page/tab) and all subcomponents in folders with a descriptive name will ease finding things

Reply:
In difference to your suggestion in previous discussion, I intentionally chose not to name subdirectory `fields/components`, instead naming it  `fields/modal_content`. I googled a bit for typical structural patterns in React application, finding blogs like [this](https://medium.com/swlh/demystifying-the-folder-structure-of-a-react-app-c60b29d90836), [this](https://www.freecodecamp.org/news/best-file-structure-for-react-components/) and [this](https://www.taniarascia.com/react-architecture-directory-structure/). From what I've seen, a `component` dir is typically reserved for a self-contained, even stateless components. This matches nicely with our `js/components` that (mostly) holds common components like `RequiredMark`. The splitting of files in this case does not have the same organizational purpose, the "add fields" modal is very much a unique and stateful component. The purpose of this split is just to increase code readability, to avoid massive files with thousands of lines of code. For such a purpose, to me a sensible place seems either a named subdirectory, like in this PR, or as a sibling to parent component, like for example [here](https://github.com/liferay/liferay-portal/tree/master/modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/actions).

The blogs of the interwebs also talk about `containers` main directory, that would hold all stateful components. In our case, that would be everything besides `js/utils` and `js/components`. This seems a bit radical structure change, and a bit too disruptive for now, but it's something we may consider. What do you think?

Reply:
It seems that there is no perfect solution for everybody, it is just a matter of finding what work for us and extract a pattern to follow. 

I like that the folders and components are named after their intent (`/modal_content` & `AddFieldsModalContent.tsx`) and this is something that I'm missing in other places of the portal, like in [Objects](https://github.com/liferay/liferay-portal/tree/master/modules/apps/object/object-web/types/src/main/resources/META-INF/resources/js/components), where there are lots of files and you never know which one corresponds with the page, the tab of any other part of the UI you need to modify.

From the last article you share, I like the idea of [views - or pages](https://www.taniarascia.com/react-architecture-directory-structure/#views) as a sibling folder of `js/components` and `js/utils`. This, combined with the idea of having all related things in the same folder, could end up with a feature (x.e. `frontend-data-set-views`) that have folders for `js/components`, `js/utils` and `js/pages`, then for each page the same division or organization could be applied. Each page could have components, utils and any sub-page or page alike component (tab).

Reply:
I like `views` a bit more than `pages`, as it is more abstract. A "page" implies to me the whole screen, while `view` can be any presentation component. Then again, that's the unfortunate bit with "view", it has strong implication for a presentation-only implantation. And we already have this, `component`. I see how some people got to `containers`. It's technically probably the most correct compromise, but it has the problem of being too confusing :sweat_smile: 

But overall, I really like the split into three parts, and the recursive structure you are suggesting! :+1:  We are planning to migrate this whole module to a standalone app. If we don't change this on the way, we can take the opportunity to reshuffle files then.

Reply:
Yep, basically having a predictable structure could help in some cases. Probably, there are lots of cases were this kind of deep hierarchy is too much.

<table>
<tr>
<td> Current </td> <td> Suggested </td>
</tr>
<tr>
<td>

```
.
└── /js
    ├── /components
    │   ├── /modal_content
    │   │   ├── ClientExtensionFilter.tsx
    │   │   ├── DateRangeFilter.tsx
    │   │   └── SelectionFilter.tsx
    │   ├── CheckboxMultiSelect.tsx
    │   ├── OrderableTable.tsx
    │   ├── RequiredMark.tsx
    │   ├── Search.tsx
    │   └── ValidationFeedback.tsx
    ├── /fds_view
    │   ├── /actions
    │   │   ├── ActionForm.tsx
    │   │   └── ActionList.tsx
    │   ├── /fields
    │   │   ├── /modal_content
    │   │   │   └── AddFieldsModalContent.tsx
    │   │   └── Fields.tsx
    │   ├── Actions.tsx
    │   ├── Details.tsx
    │   ├── Filters.tsx
    │   ├── Pagination.tsx
    │   └── Sorting.tsx
    ├── /utils
    │   ├── openDefaultFailureToast.ts
    │   └── openDefaultSuccessToast.ts
    ├── api.ts
    ├── Constants.ts
    ├── FDSEntries.ts
    ├── FDSView.ts
    ├── FDSViews.ts
    └── types.ts
```

</td>
<td>

```
.
└── /js
    ├── /components
    │   ├── CheckboxMultiSelect.tsx
    │   ├── OrderableTable.tsx
    │   ├── RequiredMark.tsx
    │   ├── Search.tsx
    │   ├── ValidationFeedback.tsx
    ├── /utils
    │   ├── openDefaultFailureToast.ts
    │   ├── openDefaultSuccessToast.ts
    ├── /views|pages|containers
    │   ├── /FDSEntries
    │   │   ├── FDSEntries.tsx
    │   │   └── /utils
    │   │       └── whatever.ts
    │   ├── /FDSView
    │   │   ├── /Actions(Tab)
    │   │   │   ├── Actions.tsx
    │   │   │   └── /components
    │   │   │   │   ├── /form
    │   │   │   │   │   └── ActionForm.tsx
    │   │   │   │   └── /list
    │   │   │   │       └── ActionList.tsx
    │   │   ├── /Details(Tab)
    │   │   │   ├── Details.tsx
    │   │   ├── /Fields(Tab)
    │   │   │   ├── Fields.tsx
    │   │   │   ├── /components
    │   │   │   │   ├── /modal_content
    │   │   │   │   │   └── AddFieldsModalContent.tsx
    │   │   ├── /Filters(Tab)
    │   │   │   ├── Filters.tsx
    │   │   │   ├── /components
    │   │   │   │   ├── /modal_content
    │   │   │   │   │   ├── ClientExtensionFilter.tsx
    │   │   │   │   │   ├── DateRangeFilter.tsx
    │   │   │   │   │   └── SelectionFilter.tsx
    │   │   ├── /Pagination(Tab)
    │   │   │   └── Pagination.tsx
    │   │   └── /Sorting(Tab)
    │   │       └── Sorting.tsx
    │   └── /FDSViews
    │            └── FDSViews.tsx
    ├── api.ts
    ├── Constants.ts
    └── types.ts
```

</td>
</tr>
</table>

Reply:
There are few things that bug me in this suggestion:
1. There is an inconsistency in format of directory names. There are:
    1. Names that match component name, in PascalCase, like `FDSEntries` and `FDSViews`.
    2. Names that are conveniently chosen, in snake_case, like `form` and `modal_content`.
    3. Brackets in directory title just look weird :smile: It may be a good option to rename components themselves to contain `*Tab` suffix, like `FiltersTab`. The problem with that is we may change implementation into something other than tabs down  the line. 
1. I like the idea of keeping `components` directory strictly for common components, of presentational and stateless nature, used in multiple places in their scope. None of the components below `/js/views` match these criteria. They should be either under `views`*.

A way to fix the `views|pages|containers` dilemma is to not to have any dir in this case. This prevents potential explosion in depth, as each child component has to to go two directories down (for example `/views/form`).

With above in mind, here's what I suggest:


```
.
└── /js
    ├── /components
    │   ├── CheckboxMultiSelect.tsx
    │   ├── OrderableTable.tsx
    │   ├── RequiredMark.tsx
    │   ├── Search.tsx
    │   └── ValidationFeedback.tsx
    ├── /utils
    │   ├── openDefaultFailureToast.ts
    │   └── openDefaultSuccessToast.ts
    ├── /fds_entries
    │   └── FDSEntries.tsx
    ├── /fds_view
    │   ├── FDSView.tsx
    │   ├── /actions
    │   │   ├── Actions.tsx
    │   │   ├── /form
    │   │   │   └── ActionForm.tsx
    │   │   ├── /list
    │   │   │    └── ActionList.tsx
    │   │   ├── /components
    │   │   │    └── MyComponent.tsx // used only in Actions, ActionForm and ActionList
    │   │   └── /utils
    │   │        └── doSomething.ts // used only in Actions, ActionForm and ActionList
    │   ├── /details
    │   │   └── Details.tsx
    │   ├── /fields
    │   │   ├── Fields.tsx
    │   │   ├── /modal_content
    │   │   │   └── AddFieldsModalContent.tsx
    │   ├── /filters
    │   │   ├── Filters.tsx
    │   │   ├── /modal_content
    │   │   │   ├── ClientExtensionFilter.tsx
    │   │   │   ├── DateRangeFilter.tsx
    │   │   │   └── SelectionFilter.tsx
    │   ├── /pagination
    │   │   └── Pagination.tsx
    │   └── /sorting
    │       └── Sorting.tsx
    ├── /fds_views
    │      └── FDSViews.tsx
    ├── api.ts
    ├── Constants.ts
    └── types.ts
```

Reply:
@dsanz @bryceosterhaus @izaera any thoughts on the subject? :brain: :zap: 

Also, @juanjofgliferay , feel free to review the rest of the PR :smile:

Reply:
I don't have a clear picture or opinion on how this should be structured, but you suggestions looks good to me :+1: .

I agree views is a bit generic, but once you know we call it FDS views, it's OK. Pages would be much worse, because Pages are a core thing of DXP so better not use it for anything, I guess.